Merge Sample Type UI creation and editing into one page

Sigmonia opened the pull request.

labkey-bpatel committed:
List Designer - initial commit:
- Configure list module for React page.
- Add Experimental feature to toggle on/off the new React page.

labkey-bpatel committed:
List Designer - initial commit:
- Configure list module for React page.
- Add Experimental feature to toggle on/off the new React page.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_list_designer_6798

labkey-bpatel committed:
Update devDependencies, include package-lock.json.

labkey-bpatel committed:
Update devDependencies, include package-lock.json.

labkey-bpatel committed:
ListDesigner.tsx entryPoint setup and initial wrapper for shared component.

labkey-bpatel committed:
ListDesigner.tsx entryPoint setup and initial wrapper for shared component.

labkey-bpatel committed:
Added comment.

labkey-bpatel committed:
Added comment.

labkey-bpatel committed:
Add constructor to ListDesigner. Update EditListDefinitionAction - update nav trail and move props code.

labkey-bpatel committed:
Add constructor to ListDesigner. Update EditListDefinitionAction - update nav trail and move props code.

labkey-bpatel committed:
Add API action to get ListDefinition.

labkey-bpatel committed:
Add API action to get ListDefinition.

labkey-bpatel committed:
Include domain when getting ListDefinition (in order to also get list fields in addition to List property).

labkey-bpatel committed:
Include domain when getting ListDefinition (in order to also get list fields in addition to List property).

labkey-bpatel committed:
Update build.gradle to include dependency on npmInstall command finishing at the platform repository level.

labkey-bpatel committed:
Update build.gradle to include dependency on npmInstall command finishing at the platform repository level.

labkey-bpatel committed:
Ignore lists' node_modules and auto-generated files.

labkey-bpatel committed:
Ignore lists' node_modules and auto-generated files.

labkey-bpatel committed:
Improved failure messages.

labkey-bpatel committed:
Improved failure messages.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Rename ListDesigner to Designer, plus related refactoring.

labkey-bpatel committed:
Rename ListDesigner to Designer, plus related refactoring.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Added new GetDomainKindAction (variant of existing GetDomainAction) which also returns domain kind specific properties; and related refactoring.

labkey-bpatel committed:
Added new GetDomainKindAction (variant of existing GetDomainAction) which also returns domain kind specific properties; and related refactoring.

RosalineP committed:
Establishes List Properties Panel. DomainId in API call currently hardwired.

RosalineP committed:
Establishes List Properties Panel. DomainId in API call currently hardwired.

labkey-bpatel committed:
Use designer instead of listDesigner.

labkey-bpatel committed:
Use designer instead of listDesigner.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Establishes componentDidMount. Includes AdvancedSettings modal (not fully wired)

RosalineP committed:
Establishes componentDidMount. Includes AdvancedSettings modal (not fully wired)

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_list_designer_6798

labkey-bpatel committed:
Rename GetDomainKindAction to GetDomainDetailsAction. Added new GetListPropertiesAction to be able to get domainId.

labkey-bpatel committed:
Rename GetDomainKindAction to GetDomainDetailsAction. Added new GetListPropertiesAction to be able to get domainId.

labkey-bpatel committed:
Updated package-lock.json

labkey-bpatel committed:
Updated package-lock.json

RosalineP committed:
Contains initial work on domainForm + ListDesignerPanels section. Uses Binal's GetListPropertiesAction and GetDomainDetails

RosalineP committed:
Contains initial work on domainForm + ListDesignerPanels section. Uses Binal's GetListPropertiesAction and GetDomainDetails

RosalineP committed:
Adds List Fields. Other work in progress

RosalineP committed:
Adds List Fields. Other work in progress

RosalineP committed:
(Forgot to bump package.json @labkey.components number)

RosalineP committed:
(Forgot to bump package.json @labkey.components number)

RosalineP committed:
Updates labkey-ui-components version

RosalineP committed:
Updates labkey-ui-components version

labkey-bpatel committed:
Return default empty list when listId is not provided or is null.

labkey-bpatel committed:
Return default empty list when listId is not provided or is null.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for CreateDomain:
- Generified DomainKind.
- Added new method getTypeClass() in DomainKind.
- Update ListDomainKind to implement new method with appropriate bean (ListDomainKindProperties).
- Update rest of DomainKind implementations to use JSONObject (instead of Map<String, Object>) so that it works with generified DomainKind (using Map<String, Object> did not work as desired with generics so resorted to using JSONObject).
(In the future when we convert other DomainKinds, use the approach taken for Lists - see ListDomainKind.getTypeClass() & createDomain()).

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for CreateDomain:
- Generified DomainKind.
- Added new method getTypeClass() in DomainKind.
- Update ListDomainKind to implement new method with appropriate bean (ListDomainKindProperties).
- Update rest of DomainKind implementations to use JSONObject (instead of Map<String, Object>) so that it works with generified DomainKind (using Map<String, Object> did not work as desired with generics so resorted to using JSONObject).
(In the future when we convert other DomainKinds, use the approach taken for Lists - see ListDomainKind.getTypeClass() & createDomain()).

labkey-bpatel committed:
Remove unused class pushed by mistake.

labkey-bpatel committed:
Remove unused class pushed by mistake.

labkey-bpatel committed:
Comment & Javadoc for getTypeClass().

labkey-bpatel committed:
Comment & Javadoc for getTypeClass().

labkey-bpatel committed:
Update comment.

labkey-bpatel committed:
Update comment.

labkey-bpatel committed:
Update comment, again.

labkey-bpatel committed:
Update comment, again.

labkey-bpatel committed:
Minor updates - return default integer enum value instead of text.

labkey-bpatel committed:
Minor updates - return default integer enum value instead of text.

labkey-bpatel committed:
Minor updates - match the field type to field type in db.

labkey-bpatel committed:
Minor updates - match the field type to field type in db.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Supports load and create list scenarios

RosalineP committed:
Supports load and create list scenarios

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for SaveDomainAction & and override updateDomain() for List:
- Change to DomainKind.updateDomain() signature to include nullable DomainKind specific properties/options & refactor callers.
- Update SaveDomain js api to include optional DomainKind specific properties/options.

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for SaveDomainAction & and override updateDomain() for List:
- Change to DomainKind.updateDomain() signature to include nullable DomainKind specific properties/options & refactor callers.
- Update SaveDomain js api to include optional DomainKind specific properties/options.

labkey-bpatel committed:
Add evaluationDependsOn()

labkey-bpatel committed:
Add evaluationDependsOn()

RosalineP committed:
Do a permissions check. Update components version

RosalineP committed:
Do a permissions check. Update components version

labkey-bpatel committed:
Update client api docs to deprecate LABKEY.Domain.get. Include its replacement - LABKEY.Domain.getDomainDetails.

labkey-bpatel committed:
Update client api docs to deprecate LABKEY.Domain.get. Include its replacement - LABKEY.Domain.getDomainDetails.

labkey-bpatel committed:
Update to client api docs.

labkey-bpatel committed:
Update to client api docs.

labkey-bpatel committed:
Added further validation. Improved exception throwing.

labkey-bpatel committed:
Added further validation. Improved exception throwing.

labkey-bpatel committed:
Remove check on JSONObject.class

labkey-bpatel committed:
Remove check on JSONObject.class

Sigmonia committed:
Initial work to reduce the two step process of creating/updating a SampleType, into one.
- Update actions
- Updated UI
- Updated SampleType DomainKinds to work with the List Designer changes

TODOs
- Fix DKs that inherit from SampleSetDomainKind
- Bump alpha components package
- Fix idCols saving/updating
- Fix error handling on panel
- Fix jest tests
- Fix SM tests
- more TBD

RosalineP committed:
Bump ui-components to newly published version

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Bump ui-components.

labkey-bpatel committed:
Use ArrayList to avoid ClassCastException error. (arguments.get("idCols") is an ArrayList and casting to Object[] was throwing ClassCastException)

labkey-bpatel committed:
Update ListDomainKindProperties not to extend Entity; and related refactoring.

RosalineP committed:
Bumping ui-components version

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

labkey-bpatel committed:
Get list via domainId and not name (since changing a name of the list doesn't update the name in exp.domainDescriptor, so using GWTDomain.getName() returns an old name).

cnathe committed:
Item 6639: minor updates to list React setup files and add a README for the module

cnathe committed:
Item 6639: List designer updates to use functions from new shared component package

cnathe committed:
Item 6639: fix issue with save not getting listId based on the container

Sigmonia committed:
in-progress recovery point
- Merged Sample Type details and domain design in LKS

cnathe committed:
Use labkey primary button styling for domain designers in LKS pages

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

cnathe committed:
Item 6639: list designer support for dirty state check on navigate
- also add query to get listId for newly saved list for navigating to list grid

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

cnathe committed:
Item 6801: For new list creation, apply options/properties to the newly created list definition in ListDomainKind.createDomain

labkey-susanh committed:
Issue 39091: Use text instead of icon for "new sample set" action button

labkey-susanh committed:
Update ui-components alpha version to get newly exported methods and properties

cnathe committed:
Item 6639: Update to latest @labkey/components alpha package version

cnathe committed:
Item 6801: Fix for duplicate name check on list designer create and update

labkey-bpatel committed:
Code Review Changes:
- Corrections to javadocs and client api docs.
- Remove DomainKindProperties interface, and related refactoring.
- Add convenience subclass of AbstractDomainKind named BaseAbstractDomainKind, and related refactoring.

cnathe committed:
Item 6801: update list grid "view design" link text for new designer (i.e. just use "design" for link text)

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Update to latest @labkey/components alpha package version after merge from master

Sigmonia committed:
Merge from fb_list_designer

Sigmonia committed:
ignore generated SampleTypeDesigner views

cnathe committed:
Update to latest @labkey/components alpha package version

labkey-bpatel committed:
Code Review change: Removed name conflicts check since its already being checked in centralized domain code. Update list properties except listId, domainId, keyName and keyType.

labkey-bpatel committed:
Code review changes:
- Handle cases when existing key field is null or is not provided in the updated domainDesign.
- Remove lastIndexed property from being updated.
- Update client api doc listing read-only props for List.
- Set empty string as default values for entireListTitleTemplate & eachItemTitleTemplate properties.

labkey-bpatel committed:
Slight rewording.

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Update to latest @labkey/components alpha package version

RosalineP committed:
Bumping component version

Sigmonia committed:
Merge from fb_list_designer

labkey-bpatel committed:
Change to isEmpty() check instead of null check.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Code review change : Remove unused generic type, and related refactor.

labkey-bpatel committed:
Code Review changes:
- Use ApiUsageException instead of IllegalArgumentException for CreateDomain.
- Add to ValidationException error obj instead of using IllegalArgumentException.
- Remove null check when updating props to allow for change from a non-null to null val (except for name).

Sigmonia committed:
Bump alpha components
Fix some SampleTypeDesigner issues
clean-up some commented out code

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6798: one more update to Domain.js docs

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Update to latest @labkey/components alpha package version

RosalineP committed:
Bumping component version

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	experiment/package-lock.json
#	experiment/package.json

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6639: handle list creation case with error during file import

RosalineP committed:
Bumping components version

RosalineP committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

# Conflicts:
#	list/package.json

RosalineP committed:
Bumping components version

labkey-susanh committed:
Merge from list_designer

RosalineP committed:
Bumping components version

Sigmonia committed:
not sure why these are missing...

cnathe committed:
Item 6798: automated test updates to prep tests that use ListHelper.createList()

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

labkey-susanh committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_mergeSTUI_6848

labkey-susanh committed:
Merge branch 'fb_mergeSTUI_6848' of https://github.com/LabKey/platform into fb_mergeSTUI_6848

labkey-susanh committed:
Update ui-components alpha version

labkey-tchad committed:
Remove redundant 'Edit Fields' button from sample set details page

Sigmonia committed:
merge from master

Sigmonia committed:
merge from master

Sigmonia committed:
Update domain apis to validate options parameter

labkey-tchad committed:
More test refactors and updates for new Sample Type UI

Sigmonia committed:
Bump alpha components version

Sigmonia committed:
Add new parameter for SampleTypeDesigner
Bump alpha components version

cnathe commented on api/src/org/labkey/api/exp/api/DomainKindProperties.java:
```diff
@@ -0,0 +1,8 @@
+/*
+** Interface for DomainKind specific properties.
+*/
+public interface DomainKindProperties
```
Have you done a recent merge from develop? Binal ended up removing this class at the end of her PR changes for the list designer platform changes. I assume this file will go away once that merge happens.

cnathe commented on api/src/org/labkey/api/exp/api/ExpSampleSet.java:
```diff
@@ -35,7 +35,7 @@
  * Material version of {@link ExpDataClass}
  */
-public interface ExpSampleSet extends ExpObject
+public interface ExpSampleSet extends ExpObject, DomainKindProperties
```
see comment from last file about this DomainKindProperties class being remove

cnathe commented on api/src/org/labkey/api/exp/api/SampleTypeDomainKindProperties.java:
```diff
@@ -0,0 +1,154 @@
+import java.util.List;
+import java.util.Map;
+
+//TODO rename? SampleTypeDomainDetails?
```
do you plan to leave this TODO or make the change?

Sigmonia replied:
removed TODO, kept name same to match with ListDomainKindProperties

cnathe commented on api/src/org/labkey/api/exp/api/SampleTypeDomainKindProperties.java:
```diff
@@ -0,0 +1,154 @@
+            this.lsid = ss.getLSID();
+            this.description = ss.getDescription();
+            this.idCols = new ArrayList<>();
+            ss.getIdCols().forEach(col -> this.idCols.add(col.getPropertyId()));  //TODO verify this...
```
leave TODO or ...?

cnathe commented on api/src/org/labkey/api/exp/api/SampleTypeDomainKindProperties.java:
```diff
@@ -0,0 +1,154 @@
+    @JsonIgnore
+    public JSONObject toJSONObject()
+    {
+        //TODO probably need to do more here... see MassSpecFractionsDomainKind.createDomain
```
leaving TODO or ...?

Sigmonia replied:
removed TODO

cnathe commented on api/src/org/labkey/api/exp/property/DomainKind.java:
```diff
@@ -299,4 +300,18 @@ public String getObjectUriColumnName()
+    public void validateOptions(Container container, User user, T options, boolean isUpdate)
+    {
+        if (!isUpdate && !this.canCreateDefinition(user, container))
+            throw new UnauthorizedException("You don't have permission to create a new domain");
```
I like this base validateOptions helper. This check does the create case, you could likely also add the update case as something like: 
`if (isUpdate && this.canEditDefinition(...)`
but looks like you might have to pass in the Domain object in that case so I can see leaving it out.

Sigmonia replied:
passed down Domain and added Update check

cnathe commented on assay/src/org/labkey/assay/PlateBasedAssaySampleSetDomainKind.java:
```diff
@@ -133,8 +137,18 @@ public boolean showDefaultValueSettings()
+    @Override
+    public ValidationException updateDomain(GWTDomain<? extends GWTPropertyDescriptor> original, GWTDomain<? extends GWTPropertyDescriptor> update, @Nullable SampleTypeDomainKindProperties options, Container container, User user, boolean includeWarnings)
+    {
+        //TODO this is probably broken too. Maybe we should throw a not supported exception here?
```
Same question from the commonAssay PR related to this. I included Josh to get his review of that PR so we can see what he says.

Sigmonia replied:
This method isn't called as far as I can tell. This overrides the underlying SampleSetDomainKind, but the implemented Assay gets updated through the AssayService. The better option here I think is throwing a NotImplementedException or similar.

cnathe commented on experiment/src/client/entryPoints.js:
```diff
@@ -9,5 +9,10 @@ module.exports = {
         path: './src/client/DomainDesigner'
+    }, {
+        name: 'sampleTypeDesigner',
+        title: 'Field Editor',
```
I would change this title to be 'Sample Type Designer'

Sigmonia replied:
Changed.

cnathe commented on experiment/src/org/labkey/experiment/XarExporter.java:
```diff
@@ -548,10 +548,11 @@ public void addSampleSet(ExpSampleSet sampleSet)
+//        if (sampleSet.getParentCol() != null)
+//        {
+//            xSampleSet.setParentField(getPropertyName(sampleSet.getParentCol()));
+//        }
```
leaving TODO or ...?

Sigmonia replied:
Uncommented for now. I don't feel confident in removing block for now.

cnathe commented on experiment/src/org/labkey/experiment/api/ExpSampleSetImpl.java:
```diff
@@ -254,6 +254,7 @@ public DomainProperty getIdCol3()
         return getDomainProperty(_object.getIdCol3());
     }
 
+    //TODO remove
```
leaving TODO or ...?

Sigmonia replied:
Removed TODO for now. I don't feel confident in removing method atm.

cnathe commented on experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java:
```diff
@@ -707,6 +701,17 @@ public ModelAndView getView(ExpObjectForm form, BindException errors)
                     deleteButton.setActionType(ActionButton.Action.LINK);
                     detailsView.getDataRegion().getButtonBar(DataRegion.MODE_DETAILS).add(deleteButton);
                 }
+                else
```
Just curious when we would ever end up in this else block? If we get to this code wouldn't the domainKind always have to be a SampleSetDomainKind?

Sigmonia replied:
@labkey-tchad I also think we can remove this block, did you run into a scenario where that wasn't the case?

cnathe commented on experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java:
```diff
@@ -707,6 +701,17 @@ public ModelAndView getView(ExpObjectForm form, BindException errors)
+                    {
+                        editURL.addParameter(ActionURL.Param.returnUrl, getViewContext().getActionURL().toString());
+                        ActionButton editTypeButton = new ActionButton(editURL, "Edit Fields");
+                        editTypeButton.setDisplayPermission(UpdatePermission.class);
```
UpdatePermission doesn't seem like it would be a high enough permissions here, but as per the other comment, I'm not sure what type of domain kinds would end up here anyway.

Sigmonia replied:
Update permission is what was used previously. But I think you're correct and we can probably just delete this block.

cnathe commented on experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java:
```diff
@@ -3480,6 +3485,7 @@ private DataRegion getMaterialSourceRegion(ViewContext model)
     }
 
     @RequiresPermission(DesignSampleSetPermission.class)
+    @Deprecated //Use PropertyController.SaveDomainAction
```
I would not have expected that this action would be deprecated. I see that you removed the UpdateMaterialSourceApiAction, I could see that one being deprecated (but I think removing it is better/correct). Isn't this UpdateMaterialSourceAction still going to be the action in the LKS UI for updating a sample type design?

Sigmonia replied:
I was thinking about the post action, forgot the view was also coming  from this. Removed deprecation marker.

cnathe commented on experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java:
```diff
@@ -3510,24 +3516,7 @@ public NavTree appendNavTrail(NavTree root)
-    }
-
-    @RequiresPermission(DesignSampleSetPermission.class)
+    @Deprecated //Use PropertyController.CreateDomainAction
```
Same comment as above. Won't this CreateSampleSetAction still be where the LKS UI goes for creating a new sample type design?

Sigmonia replied:
I was thinking about the post action, forgot the view was also coming from this. Removed deprecation marker.

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -66,7 +68,9 @@
-public class SampleSetDomainKind extends BaseAbstractDomainKind
+public class SampleSetDomainKind extends AbstractDomainKind<SampleTypeDomainKindProperties>
+//TODO Remove or use?
+//public class SampleSetDomainKind extends AbstractDomainKind<SampleTypeDomainKindProperties>
```
remove comment?

Sigmonia replied:
removed. (I feel like I've removed some of these TODOs previously)

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -257,8 +267,111 @@ public boolean canDeleteDefinition(User user, Domain domain)
+    {
+        if (options == null)
+        {
+            return;  //TODO verify this is correct...
```
wouldn't you still want to call the super.validateOptions in this case? It has a permissions check in it which still would seem valid.

Sigmonia replied:
fixed. Moved to after super call

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -257,8 +267,111 @@ public boolean canDeleteDefinition(User user, Domain domain)
         return domain.getContainer().hasPermission(user, DesignSampleSetPermission.class);
     }
 
+    //TODO Remove or use?
```
I'm guessing the answer is that you are using this and can remove this comment.

Sigmonia replied:
TODO removed.

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -257,8 +267,111 @@ public boolean canDeleteDefinition(User user, Domain domain)
+//                if (ss != null)
+//                    throw new IllegalArgumentException("A Sample Type with that name already exists.");
+//            }
+//        }
```
These seem like valid test cases for a new sample set create case. Are they commented out because they are called elsewhere and these lines can be removed? Or should they be uncommented?

Sigmonia replied:
Name is coming through the GwtDesign. Passed in as a parameter and uncommented

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -257,8 +267,111 @@ public boolean canDeleteDefinition(User user, Domain domain)
+        //TODO this supplied within the domainDesign fields
+//        int nameMax = materialSourceTI.getColumn("Name").getScale();
+//        if (name != null && name.length() >= nameMax)
+//            throw new IllegalArgumentException("Value for Name field may not exceed " + nameMax + " characters.");
```
same as comment about, this seems like a valid check that we either know is happening elsewhere and this can be deleted or that it should be uncommented.

Sigmonia replied:
Name is coming through the GwtDesign. Passed in as a parameter and uncommented

cnathe commented on internal/src/org/labkey/experiment/api/SampleSetDomainKind.java:
```diff
@@ -257,8 +267,111 @@ public boolean canDeleteDefinition(User user, Domain domain)
+
+        //TODO verify rowId value for create
+        SampleSetService ss = SampleSetService.get();
+        ExpSampleSet sampleSet = options.getRowId() != -1 ? ss.getSampleSet(options.getRowId()) : null;
```
I don't think that zero is a valid RowId either. Should this check be `options.getRowId() > 0`?

Sigmonia replied:
I thought 0 was valid, but not positive (no pun intended), but it should be > rather than !=

cnathe commented on experiment/src/client/entryPoints.js:
```diff
@@ -9,5 +9,10 @@ module.exports = {
+    }, {
+        name: 'sampleTypeDesigner',
+        title: 'Field Editor',
+        permission: 'admin', // this is admin so that direct access to this view has highest level perm, see PropertyController.EditDomainAction for main usage
```
This permissions of 'admin' is correct, but the comment refers to PropertyController. EditDomainAction, I think that should actually be ExperimentController.BaseSampleSetAction right?

Sigmonia replied:
corrected.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+}
+
+//TODO should these be moved to a constants file? or shared through components?
+export const NAME_EXPRESSION_TOPIC = 'sampleIDs#expression';
```
This const exists on the shared component side and I was not expecting that it would be needed here. I would have thought that it could be set in the SampleTypeDesigner shared component and not have to be used / passed as a prop here.

Sigmonia replied:
removed.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+    lookupQuery: 'Materials',
+    lookupType: {...SAMPLE_TYPE},
+    name: 'SampleId',
+} as Partial<IDomainField>;
```
I haven't looked at the PR on the labkey-ui-components side yet, but how is this field config getting used? Do we end up adding a default first field in the SampleTypeDesigner component or something like that? I would have expected the SampleTypeDesigner fields domain to start out empty in this case.

Sigmonia replied:
This is used to match previous behavior which adds a (hidden from designer) SampleId field. I think it should probably be visible and readOnly but was trying to match original behavior first.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+    constructor(props) {
+        super(props);
+
+        initQueryGridState();
```
For my implementation of the DataClassDesigner shared component, I put this `initQueryGridState()` call inside of the DataClassPropertiesPanel constructor and it looks to be working as expected and I didn't have to do the same call here in the LK usage side.

Sigmonia replied:
Moved to ui-components SampleTypeDesigner constructor

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+        const { RowId, schemaName, queryName, returnUrl } = ActionURL.getParameters();
+
+        let messages = List<IBannerMessage>();
+        if ((!schemaName || !queryName) && !RowId) {
```
Since the LK button that generates the URL for the Sample Type designer edit case only uses RowId as a param, I don't think we need to also support the schemaName/queryName params. I would drop that. Also, since this component is used in both the create and update case. I wasn't expecting a check here in the constructor. Instead I thought we'd skip the check here and then conditionally, in the componentDidMount either load the existing one if there is a rowId and assume it is the create case if not.

Sigmonia replied:
Added check for action, and banner message if update case and no rowId parameter.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+        const { domainId, schemaName, queryName, rowId, messages } = this.state;
+
+        if ((schemaName && queryName) || domainId ) {
+            this.fetchSampleTypeDomain(domainId, schemaName, queryName);
```
Is this case needed? won't the only way to get to this component be either for the create case (which doesn't have a rowId param) or for the update case (with a rowId param)?

Sigmonia replied:
I think the edit fields button was supplying schema and query, but since we removed that I think it is safe to just target the rowId.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+            this.fetchSampleTypeDomain(domainId, schemaName, queryName);
+        }
+        else if (rowId) {
+            getSampleSet({rowId})
```
Would it make sense to put this as an action/helper on the shared component side? I haven't looked at the PR for sapmleManagement yet, but assume it might have to do something similar to this?

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+    /**
+     * Look up full Sample Type domain, including fields
+     **/
+    private fetchSampleTypeDomain = (domainId, schemaName?, queryName?): void => {
```
Would it make sense to put this as an action/helper on the shared component side? I haven't looked at the PR for sapmleManagement yet, but assume it might have to do something similar to this?

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+        const {sampleType} = this.state;
+        const subtitle = 'Edit Sample Type Details';
+
+        //TODO check permissions
```
I don't think we need this permissions check on the LKS side. The two ways to view this component would be: 1) through the ExperimentController actions which already have the correct permissions check 2) via the experiment-sampleTypeDesigner.view resource/view, which via the entryPoint.js file is set as admin perm required.

Sigmonia replied:
removed.

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,225 @@
+        //     return <InsufficientPermissionsPage title={subtitle}/>
+        // }
+        // else
+        if (menuLoading) {
```
I don't see menuLoading in the props type definition. Also, I don't think we would be waiting for any menu to load here in the LKS side (I assume that part came from the SM app code?). Maybe calling this isLoading, but I assume it is a state variable not prop?

Sigmonia replied:
removed. Never set or tracked.

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

Sigmonia committed:
Remove DomainKindProperties interface, since merge did not.

Sigmonia committed:
Remove TODOs and fix args NPE

Sigmonia committed:
Allow using DataClasses in SampleTypeDesigner
fix error messages

Sigmonia committed:
bump alpha versions

Sigmonia committed:
Bump alpha version of components

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

Sigmonia committed:
CR feedback changes
- Added canUpdate check to DomainKind.validateOptions and some additional parameters
- Fixed comments
- Removed TODOs
- Added some additional validations checks to SampleSetDomainKind.validateOptions

labkey-tchad committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

Sigmonia committed:
CR feedback changes
-

Sigmonia committed:
Merge remote-tracking branch 'origin/fb_mergeSTUI_6848' into fb_mergeSTUI_6848

Sigmonia committed:
Issue 39874: Parent aliases with the same label should show error

Sigmonia committed:
bump alpha versions

cnathe commented on experiment/src/client/SampleTypeDesigner/SampleTypeDesigner.tsx:
```diff
@@ -0,0 +1,220 @@
+
+        this.state = {
+            schemaName,
+            queryName,
```
It doesn't look like schemaName or queryName are expected anymore or used in this component, so I believe they can be removed from the state interface. Looks like fetchSampleTypeDomain is only using the domainId now.

Sigmonia committed:
bump alpha version

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

Sigmonia committed:
Alpha version update after merge from master/develop

labkey-tchad committed:
Update tests for new Sample Type designer

labkey-tchad committed:
Merge branch 'develop' into fb_mergeSTUI_6848

labkey-tchad commented:
I opened an [issue](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=39987) to track problems arising from this refactor.

Sigmonia committed:
Merge from develop

Sigmonia committed:
Update package-locks

cnathe committed:
Issue 39995: CreateDomainAction form validation of options does not apply if being created from a template

labkey-susanh committed:
Issue 39993: Allow for supplying a read-only name when creating a new sample type

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

cnathe committed:
Update to latest @labkey/components package alpha version

cnathe committed:
Fix test failure related to transaction commit issue when there are domain update errors
- DomainDesignerTest.testUserWarningOnRequiredFieldWithEmptyValues failed when trying to verify error message from setting a required field that has empty values in the domain

cnathe committed:
Update to latest @labkey/components alpha package version

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_mergeSTUI_6848

labkey-susanh committed:
Merge from develop and update ui-components alpha version

cnathe committed:
SampleTypeDesigner fix for validation of parent alias header against the domain fields property names
- SampleSetParentColumnTest.testAliasNameConflictsWithFieldName fix

Sigmonia committed:
Bump alpha version .23

Sigmonia committed:
Update to release version of components to v0.37.0
